src/core/shape_inference/include/broadcast_base_shape_inference.hpp

```diff
@@ -0,0 +1,303 @@
+        } else {
+            NODE_VALIDATION_CHECK(op,
+                                  arg_shape_vec[i].compatible(target_shape[i]),
+                                  "Broadcast incorrect target shape. Expecting either 1 or ",
+                                  arg_shape_vec[i],
+                                  ". Got ",
+                                  target_shape[i]);
+            result_shape[i] = target_shape[i];
+        }
+    }
+}
+
+template <class T>
+void broadcase_base_shape_infer(const ov::op::util::BroadcastBase* op,
+                                const std::vector<T>& input_shapes,
+                                std::vector<T>& output_shapes,
+                                const std::map<size_t, std::shared_ptr<ngraph::runtime::HostTensor>>& constant_data) {
+    using DimType = typename std::iterator_traits<typename T::iterator>::value_type;
+
+    // shape node should produce a one dimensional shape.
+    auto broadcast_shape_rank = input_shapes[1].rank();
+    NODE_VALIDATION_CHECK(op,
+                          broadcast_shape_rank.compatible(1),
+                          "Broadcast shape rank must be 1, but has ",
+                          broadcast_shape_rank);
+
+    if (op->m_mode.m_type == BroadcastType::NONE) {
+        // axes_mapping node should produce a one dimensional shape.
+        auto axes_shape_rank = input_shapes[2].rank();
+        NODE_VALIDATION_CHECK(op,
+                              axes_shape_rank.compatible(1),
+                              "Broadcast axes rank must be 1, but has ",
+                              axes_shape_rank);
+    }
+
+    auto& result_shape = output_shapes[0];
+    const auto& input_shape = input_shapes[0];
+    const auto& target_shape = input_shapes[1];
+    const bool is_target_shape_known = target_shape.is_static();
+
+    T output_shape;
+    bool output_shape_defined = get_data_as_shape<T>(1, op, output_shape, constant_data);
+
+    if (auto concat = ov::as_type_ptr<ngraph::op::v0::Concat>(op->input_value(1).get_node_shared_ptr())) {
+        auto concat_inputs = concat->inputs();
+
+        if (!output_shape_defined && concat->get_output_partial_shape(0).is_static() &&
+            concat->get_shape().size() == 1 && concat_inputs.size() == shape_size(concat->get_shape())) {
+            auto output_partial_shape = std::vector<DimType>{};
+            for (const auto& concat_input : concat_inputs) {
+                auto source_node_ptr = concat_input.get_source_output().get_node_shared_ptr();
+                if (auto source_const_ptr = ov::as_type_ptr<ngraph::op::v0::Constant>(source_node_ptr)) {
+                    output_partial_shape.emplace_back(source_const_ptr->get_axis_vector_val()[0]);
+                } else {
+                    output_partial_shape.push_back(Dimension::dynamic());
+                }
+            }
+            output_shape_defined = true;
+            output_shape = T(output_partial_shape);
+        }
+    }
```

Comment:
Because get_data_as_shape cannot cover the case that one of concat input is parameter: [this case](https://github.com/openvinotoolkit/openvino/blob/master/src/core/tests/type_prop/broadcast.cpp#L59-L77). If we handle concat with this specific way, we can get concat output shape {?, 16, 50, 50}.

Reply:
Ok, can we call this code only in case when we didn't get output shape on previous step (`output_shape_defined` is false)?

Reply:
Yes, changed.